Stabilize `lint_reasons` in rustc (RFC 2383) (#99063)
rust-lang/rust: src/doc/rustc/src/lints/levels.md

These lints exist mostly to be manually turned on via configuration, as we'll
talk about later in this section.

## expect

Sometimes, it can be helpful to suppress lints, but at the same time ensure that

> the code in question still emits them. The 'expect' level does exactly this, with
> it a lint can be expected. If the lint in question is not an emitted, a new lint
> `unfulfilled_lint_expectation` at the attribute notifying you that the expectation,
> is no longer fulfilled.

Reviewer: ```suggestion
the code in question still emits them. The 'expect' level does exactly this. If the lint in question is not  emitted, the
`unfulfilled_lint_expectation` lint triggers on the `expect` attribute, notifying you that the expectation
is no longer fulfilled.
```

this suggestion is not at all formatted.

```rust
fn main() {
    #[expect(unused_variables)]
    let unused = "Everyone ignores me";

    #[expect(unused_variables)]

>     let used = "I'm usefull";

Reviewer: ```suggestion
    let used = "I'm useful";
```

Author: Me and my spelling, a never-ending story. Soon I'll have finished season two, you can preorder a copy if you'd like xD

    println!("The `used` value is equal to: {:?}", used);
}
```

This will produce the following warning:

```txt
warning: this lint expectation is unfulfilled
 --> src/main.rs:7:14
  |
7 |     #[expect(unused_variables)]
  |              ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

This level can only be defined via the `#[expect]` attribute and not via the
console. Lints with the special 'force-warn' lint will still be emitted as usual.

> The fulfillment of expectations is still tracked.

Reviewer: Not sure if this is an implementation detail. I don't think the average Rust user has any idea what this sentence means.

Author: Might be, should I remove the sentence completely or rephrase it? It could say: "The lint suppression can be overridden by the `--force-warn` argument. The lint will then be emitted as a warning, but the expectation will still be marked as fulfilled if a lint was emitted."
